Multithreading todo parsing in TodoVisitor.

criske opened the pull request:
Just in case we'll need a multithreading aproach to parse todos.

criske committed:
Multithreading todo parsing in TodoVisitor.

charlesmike commented:
@criske thank you for your Pull Request. I'll assign someone to review it soon.

criske commented on src/main/java/com/selfxdsd/todocli/TodoParser.java:
```diff
@@ -77,13 +73,16 @@ public TodoParser() {
     public List<Todo> parse(final String path) throws IOException {
         List<String> lines = Files.readAllLines(Paths.get(path));
         List<Todo> todos = new ArrayList<>();
+        final AtomicInteger currentIndex = new AtomicInteger();
```
@amihaiemil I've used a here an atomic int to be able to pass currentIndex to `parseMultiLineTodo` as a mutable ref. Before `currentIndex` was a class member and thus not thread safe.

criske commented on src/main/java/com/selfxdsd/todocli/TodoVisitor.java:
```diff
@@ -102,25 +116,33 @@ public FileVisitResult visitFile(
+                        log(todo.toString() + suffix);
+                    }
+
+                    this.serializer.addAll(todos);
```
@amihaiemil the serializer impl is already thread safe.

charlesmike commented:
@amihaiemil I couldn't find any assignee for this task. This is either because there are no contributors with role ``REV`` available or because the project does not have enough funds.

Please, make sure there is at least one available contributor with the required role and the project can afford to pay them.

amihaiemil commented on src/main/java/com/selfxdsd/todocli/TodoParser.java:
```diff
@@ -77,13 +73,16 @@ public TodoParser() {
 
-        for (currentIndex = 0; currentIndex < lines.size(); currentIndex++) {
-            String line = lines.get(currentIndex);
+        for (; currentIndex.get() < lines.size();
```
@criske This ``for`` looks really strange. Can we use a ``while`` instead?

amihaiemil commented on src/main/java/com/selfxdsd/todocli/TodoVisitor.java:
```diff
@@ -89,8 +97,14 @@ public FileVisitResult postVisitDirectory(
+            try {
+                // scanning root has finished.
+                service.shutdown();
+                service.awaitTermination(5, TimeUnit.MINUTES);
```
@criske Why do we wait for 5 minutes here?

criske replied:
That's the timeout period, basically it says the service should finish its running and queued jobs within 5 minutes. Most likely we'll never reach that timeout.

amihaiemil commented on src/main/java/com/selfxdsd/todocli/TodoVisitor.java:
```diff
@@ -102,25 +116,33 @@ public FileVisitResult visitFile(
-                    suffix = "\n";
-                } else {
-                    suffix = "";
+            service.submit(() -> {
```
@criske So, for each source code file, we start a separate thread?

criske replied:
Nope, as per service config, at most 3 threads, rest of the files will be queued.
`ExecutorService service = Executors.newFixedThreadPool(3)`

amihaiemil reviewed (changes requested):
@criske see questions :D

criske committed:
Multithreading todo parsing in TodoVisitor.

- added a flag to prevent serializing when there are errors during scanning for todos in src files.

criske commented on src/main/java/com/selfxdsd/todocli/TodoVisitor.java:
```diff
@@ -89,8 +105,19 @@ public FileVisitResult postVisitDirectory(
+                    throw new IOException("Could not serialize todos; something"
+                        + " went wrong while parsing a file for todos.");
+                } else {
+                    this.serializer.serialize();
```
@amihaiemil now serialization is done only if there are no errors while parsing the src files

amihaiemil replied:
@criske no, please remove this and we'll do it in separate ticket if necessary. I want the parsing to take place even if there is an error in one of the todos. 

That's one of the things I don't like about the original [pdd](https://github.com/yegor256/pdd) -- it crashes and reports no tickets if it finds an error, even if only 1 todo is wrong.

criske committed:
Multithreading todo parsing in TodoVisitor.

- removed the error flag, now the service won't be shutdown if there is an error while parsing a file.

amihaiemil commented:
@rultor merge it

rultor commented:
> @rultor merge it

@amihaiemil OK, I'll try to merge now. You can check the progress of the merge [here](https://www.rultor.com/t/23747-733528273)

rultor commented:
> @rultor merge it

@amihaiemil Done! FYI, the full log is [here](https://www.rultor.com/t/23747-733528273) (took me 2min)

amihaiemil commented:
@criske thanks! Paid with transaction id 96J5092377312730E
